File: src/endpoint/Bc.Contracts.Internals.Endpoint/GitHubPullRequestStatusVerification.cs
using Bc.Contracts.Internals.Endpoint.ITOps.GitHub;

namespace Bc.Contracts.Internals.Endpoint
{
    namespace Messages
    {
        public class RequestCheckPullRequestStatus
        {
            public RequestCheckPullRequestStatus(string pullRequestUri, string etag)
            {
                this.PullRequestUri = pullRequestUri;
                this.Etag = etag;
            }
        
            public string PullRequestUri { get; }

            public string Etag { get; }
        }
        
        public class ResponseCheckPullRequestStatus
        {
            public ResponseCheckPullRequestStatus(PullRequestStatus pullRequestStatus, string etag)
            {
                this.PullRequestStatus = pullRequestStatus;
                this.ETag = etag;
            }

            public PullRequestStatus PullRequestStatus { get; }

            public string ETag { get;  }
        }
    }
}

File: src/endpoint-prev/Bc.Logic.Endpoint/ITOps/GitHubPullRequestStatusVerification.fs
namespace Bc.Logic.Endpoint.ITOps.GitHubPullRequestStatusVerification

open Bc.Contracts.Internals.Endpoint.ITOps.GitHub
open Bc.Contracts.Internals.Endpoint.ITOps.GitHub.PullRequestStatusVerification
open Bc.Contracts.Internals.Endpoint.ITOps.GitHub.PullRequestStatusVerification.Messages

type PolicyLogic(configurationProvider: IConfigurationProvider) =
    member this.ConfigurationProvider = configurationProvider
    interface IPolicyLogic with
        member this.CheckPullRequestStatus(pullRequestUri, etag) =
            async {
                let! isOpenResult = GitHubApi.IsPullRequestOpen.execute
                                        GitHubConfigurationProvider.userAgent
                                        GitHubConfigurationProvider.authorizationToken
                                        pullRequestUri
                                        (Some etag)

                if isOpenResult.isOpen then
                    return ResponseCheckPullRequestStatus(PullRequestStatus.Open, isOpenResult.etag)
                else
                    let! isMerged = GitHubApi.IsPullRequestMerged.execute
                                            GitHubConfigurationProvider.userAgent
                                            GitHubConfigurationProvider.authorizationToken
                                            pullRequestUri
                    if isMerged then
                        return ResponseCheckPullRequestStatus(PullRequestStatus.Merged, isOpenResult.etag)
                    else
                        return ResponseCheckPullRequestStatus(PullRequestStatus.Closed, isOpenResult.etag)
            } |> Async.StartAsTask
            